Introduce forked upgrade testing

maurelian opened the pull request:
This PR enable running the foundry test suite against a fork of mainnet or sepolia. 

**Key changes:** 

1. A new `Upgrade.s.sol` which can be etched in place of `Deploy.s.sol`, and instead of deploying new contracts, it retrieves them from the forked node, and saves them to the deployment artifact so that they will be tested. 
2. Setup.sol has new functionality to enable fork testing: 
	-  a new state var `isForkTest` and function `skipIfForkTest()` these can be used to identify when in a fork, and skip the test or take other action if necessary. 
	- `l2Genesis.setUp()` is skipped and `Setup.L2()` returns early when in a fork test.
2. A new `just test-upgrade` command which sets the env vars necessary to run the test on a fork. This command **runs only tests in the `test/dispute` and `test/L1` directory**. This is necessary in order to keep the run time reasonable, and the majority of other tests are not relevant to forked upgrade testing.

**Limitations:**

1. At this time, many tests are still failing: [`Encountered a total of 56 failing tests, 164 tests succeeded`](https://app.circleci.com/pipelines/github/ethereum-optimism/optimism/74753/workflows/43ca7a31-b0e3-49b2-9fab-9b083cbbe783/jobs/3047492?invite=true#step-112-64754_1189). 
  - While I am working through them with good momentum, I estimate that there are 5 to 20 tweaks still needed to get them working, either by skipping them (with justifications for doing so), or making fixes to the setup, cheats and assertions.  
2. An unavoidable challenge is that it will become more difficult to write new tests as they will need to run against both locally deployed and mainnet forked systems. I believe that is a cost we have to bear, and that overtime we will become better at doing so.
3. The upgrade path itself has not yet been written, so some tests should be expected to fail for now. In the case where the failure is due to a difference between the system on mainnet vs. `develop`, they will be skipped, and labelled with `TODO(opcm upgrades):` so that they can be fixed in a child PR.

maurelian committed:
feat: Basic upgrade test scaffolding

maurelian committed:
Add Superchain Registry

maurelian committed:
Fork from Registry

maurelian committed:
feat: allowCheatcodes on setPreinstalls in L2Genesis

maurelian committed:
feat: Refactory and clean up registry reads

maurelian committed:
feat: cleanup Upgrade.s.sol

maurelian committed:
fix: Move strings into function body

Upgrade has no constructor

maurelian committed:
feat: Some cleanup

maurelian committed:
feat: write optimismPortal2

maurelian committed:
feat: Automatically run anvil on 8546

maurelian commented on .circleci/config.yml:
```diff
@@ -714,6 +714,95 @@ jobs:
             - "/root/.cache/go-build"
       - notify-failures-on-develop
 
+  contracts-bedrock-tests-upgrade:
```
I'm sure this could be done more elegantly, but it just copied and pasted from the normal `contracts-bedrock-tests` job for now to get something running in CI

maurelian committed:
feat: Add upgrade testing to Ci

maurelian committed:
fix: use parameter to set UPGRADE_TEST env var

mds1 commented on packages/contracts-bedrock/test/setup/Upgrade.s.sol:
```diff
@@ -0,0 +1,104 @@
+
+    /// @notice Reads a standard chains addresses from the superchain-registry and saves them to disk.
+    function run() public {
+        string memory superchainBasePath = "./lib/superchain-registry/superchain/configs/";
```
Could add a CI job to make sure superchain registry commit is up to date

maurelian replied:
I thought about this more since we spoke, and I'm not sure we should break the monorepo CI everytime a commit is made to the superchain registry. That would cause issues in the merge queue for one thing.

mds1 replied:
A job that runs every few hours and tags us in `#notify-ci-failures` on slack seems like a good way to account for that

maurelian replied:
I created an issue to track that.

mds1 commented on packages/contracts-bedrock/test/setup/Upgrade.s.sol:
```diff
@@ -0,0 +1,104 @@
+    function run() public {
+        string memory superchainBasePath = "./lib/superchain-registry/superchain/configs/";
+        string memory forkBaseChain = "mainnet";
+        string memory forkOpChain = "op";
```
Should these be `envOr`s so we can more easily change them in CI, e.g. to loop through many chains?

maurelian replied:
yes, done.

mds1 commented on packages/contracts-bedrock/test/setup/Upgrade.s.sol:
```diff
@@ -0,0 +1,104 @@
+    function saveProxyAndImpl(string memory _contractName, string memory _tomlPath, string memory _tomlKey) internal {
+        address proxy = vm.parseTomlAddress(_tomlPath, _tomlKey);
+        save(string.concat(_contractName, "Proxy"), proxy);
+        save(_contractName, address(uint160(uint256(vm.load(proxy, Constants.PROXY_IMPLEMENTATION_ADDRESS)))));
```
Should we verify that the implementation is not the zero address? Same for the other `saveProxyAndImpl`

maurelian replied:
sure, done.

mds1 commented on packages/contracts-bedrock/justfile:
```diff
@@ -45,6 +45,20 @@ clean:
+  anvil --silent --fork-url $MAINNET_RPC_URL --port 8546
+
+# Stops the local anvil node
+anvil-stop:
```
nit for consistency
```suggestion
anvil-fork-stop:
```

maurelian committed:
feat: more cleanup

maurelian committed:
fix: PR feedback

maurelian committed:
tweak ci

maurelian committed:
fix: justfile bug

maurelian committed:
fix: Upgrade syntax errors

maurelian committed:
add export MAINNET_RPC_URL to ci

maurelian committed:
debug address in use

maurelian committed:
feat: Allow cheatcodes for FFIInterface

tynes commented on packages/contracts-bedrock/justfile:
```diff
@@ -45,6 +45,18 @@ clean:
   forge test {{ARGS}}
 
+# Starts a local anvil node with a mainnet fork and sends it to the background
+anvil-fork:
```
nit: `ETH_RPC_URL` is much more standard, I don't think we should add non standard names, also why is this coupled to mainnet? It should be able to fork any network

tynes replied:
It would be nice to support optionally passing thru a block number to fork from

maurelian replied:
Will use ETH_RPC_URL. 

Created an [issue](https://github.com/ethereum-optimism/optimism/issues/13339) for the block number option.

tynes commented on packages/contracts-bedrock/test/setup/Upgrade.s.sol:
```diff
@@ -0,0 +1,107 @@
+contract Upgrade is Deployer {
+    using stdJson for string;
+
+    /// @notice Returns the base chain name to use for forking
```
What is a base chain?

maurelian replied:
The L1 for an L2, the L2 for an L3 etc. 
I don't think is is an entirely novel usage ([ex](https://github.com/ethereum-optimism/specs/blob/f3262e1f3b4310c39de26c50e4f076d30563b0c5/specs/protocol/isthmus/configurability.md#L45)).

tynes commented on packages/contracts-bedrock/test/setup/Upgrade.s.sol:
```diff
@@ -0,0 +1,107 @@
+
+    /// @notice Returns the OP chain name to use for forking
+    /// @return The OP chain name as a string
+    function upgradeOpChain() internal view returns (string memory) {
```
Is this the L2? it could be an L3?

maurelian replied:
We currently only support forking from chains in the superchain registry, which are limited to mainnet and sepolia, so it can only be an L2.

tynes commented on packages/contracts-bedrock/test/setup/Upgrade.s.sol:
```diff
@@ -0,0 +1,107 @@
+    function run() public {
+        string memory superchainBasePath = "./lib/superchain-registry/superchain/configs/";
+
+        // Read the superchain config files
```
Can you inline instructions on how to add a new contract?

maurelian replied:
Added this: 

```
// During development of an upgrade which adds a new contract, the contract will not yet be present in the
        // superchain-registry. In this case, the contract will be deployed by the upgrade process, and will need to
        // be saved by after the call to opcm.upgrade().
        // After the upgrade is complete, the superchain-registry will be updated and the contract will be present.
        // At this point, the test will need to be updated to read the new contract from the superchain-registry using
        // either the `saveProxyAndImpl` or `save` functions.
```

ElliotFriedman replied:
if you use fps in this repo you can write 1 proposal that will have the same code both before and after the superchain registry is updated.

1. specify deployment script + governance proposal  that upgrades contract. deployment script checks if contract is already deployed, if it is, then nothing happens in the deploy function, if it is not, then it deploys the contract. [example](https://github.com/solidity-labs-io/fps-example-repo/blob/main/src/proposals/simple-vault-multisig/MultisigProposal_01.sol#L32-L42)
2. have the Addresses.sol object read in the superchain registry as the source of truth for all addresses

maurelian replied:
introducing a new dep will expand scope too much. Open to this in the future, though a thorough issue write up would be appreciated.

tynes commented:
Do you mind describing the problem that this is trying to solve?

tynes commented on packages/contracts-bedrock/test/setup/Setup.sol:
```diff
@@ -120,9 +121,18 @@ contract Setup {
     function setUp() public virtual {
         console.log("L1 setup start!");
-        vm.etch(address(deploy), vm.getDeployedCode("Deploy.s.sol:Deploy"));
+        if (vm.envOr("UPGRADE_TEST", false)) {
```
It would be good to add some comments here about what is happening so that its easier to onboard future maintainers

tynes replied:
I see there are some comments in `Upgrade.s.sol`, those are nice

tynes commented on packages/contracts-bedrock/test/setup/Upgrade.s.sol:
```diff
@@ -0,0 +1,107 @@
+    function saveProxyAndImpl(string memory _contractName, string memory _tomlPath, string memory _tomlKey) internal {
+        address proxy = vm.parseTomlAddress(_tomlPath, _tomlKey);
+        save(string.concat(_contractName, "Proxy"), proxy);
+        address impl = address(uint160(uint256(vm.load(proxy, Constants.PROXY_IMPLEMENTATION_ADDRESS))));
```
nit: there is a 1967 helper library that can be used here but no big deal if not

maurelian replied:
oh that's nice to avoid all this casting. thx

maurelian committed:
feat: use EIP1967Helper

maurelian committed:
feat: improve comments in Setup.sol

maurelian committed:
feat: add support for interop and optimized test setups

- Implemented the ability to use interop in test configurations.
- Added checks to skip tests when certain conditions for forked tests are met.
- Enhanced Setup with options for overriding L2 genesis and creating forks.
- Included additional logging for L1 setup processes.
- Enabled etching of deployed upgrade contracts for testing purposes.

maurelian committed:
feat: Use ETH_RPC_URL as env var name

maurelian committed:
test fixes for forking

maurelian committed:
feat: anvil-fork-stop does not error

maurelian committed:
feat: Do not mess with EVM env on a fork test

maurelian committed:
fix(ctb): test fixes for a forked env

maurelian committed:
feat: Take anvil out of the loop

maurelian committed:
feat: Don't warp if forked

maurelian committed:
fix: isForkTest not forkBlock

maurelian committed:
lint

maurelian committed:
Revert "feat: Take anvil out of the loop"

This reverts commit c70e8d7fd6f65319050b06df4eeb78461a682260.

maurelian commented on .circleci/config.yml:
```diff
@@ -669,9 +673,16 @@ jobs:
-            forge <<parameters.test_command>> <<parameters.test_flags>> --match-path "$MATCH_PATH"
+            if [[ "$UPGRADE_TEST" == "true" ]]; then
+              echo "Running upgrade tests"
+              export ETH_RPC_URL="https://ci-mainnet-l1-archive.optimism.io"
```
^ getting a [429 in CI](https://app.circleci.com/pipelines/github/ethereum-optimism/optimism/74637/workflows/b5424517-0830-41fc-aea9-65ee3f0ba7fd/jobs/3043977?invite=true#step-111-102757_18) on this URL

maurelian committed:
feat: enforce forking from sepolia or mainnet

maurelian committed:
feat: remove anvil-fork-stop

also add helpful comments on just commands

maurelian committed:
fix: unsilence anvil

maurelian committed:
fix snapshot and semgrep

maurelian committed:
temp: slightly faster just-upgrade loop

maurelian committed:
feat: pin fork block

maurelian committed:
speculative: skip l2 work

maurelian committed:
feat: skip benchmark tests if isFork

maurelian committed:
feat: skip deploying L2 tokens when forking

maurelian committed:
feat: add skipIfForkTest

maurelian commented:
* **#13323** <a href="https://app.graphite.dev/github/pr/ethereum-optimism/optimism/13323?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>: 3  dependent PRs ([#13375](https://github.com/ethereum-optimism/optimism/pull/13375) <a href="https://app.graphite.dev/github/pr/ethereum-optimism/optimism/13375?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>, [#13424](https://github.com/ethereum-optimism/optimism/pull/13424) <a href="https://app.graphite.dev/github/pr/ethereum-optimism/optimism/13424?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>, [#13471](https://github.com/ethereum-optimism/optimism/pull/13471) <a href="https://app.graphite.dev/github/pr/ethereum-optimism/optimism/13471?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>) 👈 <a href="https://app.graphite.dev/github/pr/ethereum-optimism/optimism/13323?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* `develop`


<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.
<!-- Current dependencies on/for this PR: -->

maurelian committed:
feat: change upgrade_test to test_upgrade in config.yml for consistency

maurelian committed:
feat: cache forked state

maurelian committed:
feat: limit upgrade testing to L1 contracts

maurelian committed:
feat: move skipping logic into Setup.sol

maurelian committed:
feat: Simplify ci config for upgrade tests

maurelian committed:
gas snapshot

maurelian commented on .circleci/config.yml:
```diff
@@ -662,14 +666,27 @@ jobs:
+            if [[ "$(<<parameters.test_upgrade>>)" == "true" ]]; then
+              echo "Running upgrade tests"
+              export ETH_RPC_URL="https://ci-mainnet-l1-archive.optimism.io"
+              just test-upgrade
```
Unlike the non-upgrade testing path, here we use the `just` command directly, which does its own filtering to limit tests to L1 contracts.

This makes sense because it is limiting the test suite to what is relevant for fork testing. 
However it also limits our capacity to use `circleci tests split` to speed up this job in CI, which is currently very slow. 

Therefore, in order to make this workable, we need to find a way to split the tests by runtime here as well. Looking for guidance from @smartcontracts on the best way to go about that.

smartcontracts replied:
We don't use test splitting anymore anyway because it all runs on one box. I'd recommend just making this part of a new job.

maurelian committed:
delete ignored deploy artifact

maurelian committed:
remove unused added vm call

maurelian commented on packages/contracts-bedrock/test/setup/CommonTest.sol:
```diff
@@ -42,7 +43,6 @@ contract CommonTest is Test, Setup, Events {
     IOptimismMintableERC20Full L2Token;
     ILegacyMintableERC20Full LegacyL2Token;
     ERC20 NativeL2Token;
-    ERC20 BadL2Token;
```
This was unused.

maurelian committed:
feat: remove modifications to skipped test paths

maurelian committed:
feat: improve block number handling

maurelian committed:
feat: Document adding a new contract

maurelian committed:
fix: ci config syntax error

maurelian committed:
better justfile comments

maurelian committed:
feat: skip some tests and log messages when you do

maurelian committed:
feat: skip SystemConfig tests with incompatible iface

maurelian committed:
feat: add returnIfForkTest and fix a bunch of tests

maurelian committed:
feat: more test fixes and tweaks

maurelian committed:
gas snapshot

maurelian committed:
feat: Add test-upgrade-rerun to justfile and ci config

maurelian commented on .circleci/config.yml:
In this file, I elected to reuse the `contracts-bedrock-tests` job for the forked upgrade test path, rather than duplicating most of the config into a new job. Otherwise there would be a lot of redundant code, needing to be kept in sync between the two jobs.

Unfortunately, this means that there are several new points of branching logic within this job. Would appreciate an opinion from @smartcontracts on this approach vs. adding a new job for the forked upgrade.

maurelian committed:
feat: Make PINNED_BLOCK_NUMBER be not a magic number

maurelian committed:
fix: long standing bug unrelated to this work

the L1 xDM test was exercising the L2 xDM. Fork testing caught this
because it skips deploying L2 contracts.
Deploying the L2 contracts to the same network as the L1 contracts was
always a bit janky anyways.

maurelian committed:
feat: PDG: cleaner fork test handling

maurelian committed:
debugging with annotations

mds1 commented on .gitmodules:
> A new `just test-upgrade` command which sets the env vars necessary to run the test on a fork. This command **runs only tests in the `test/dispute` and `test/L1` directory**. This is necessary in order to keep the run time reasonable, and the majority of other tests are not relevant to forked upgrade testing.

Should we have a tracking issue to revisit the set of tests that are run? I ask because the sentence says "the majority of other tests are not relevant" and not "all other tests are not relevant", so perhaps we can increase coverage further

maurelian replied:
good read :)

Yes, the main ones that comes to mind is the `PreimageOracle` and other in `test/cannon`. Since we're going to track that as an action item, I may also take the opportunity to further limit the scope of tests we attempt to run with this work, so that we can come back to the ones that are trickiest to get working. Ultimately having any forked upgrade tests run is a big improvement, so I don't think we should let perfect be the enemy of good here.

maurelian replied:
Tracked: https://github.com/ethereum-optimism/optimism/issues/13411

mds1 commented on packages/contracts-bedrock/foundry.toml:
> An unavoidable challenge is that it will become more difficult to write new tests as they will need to run against both locally deployed and mainnet forked systems. I believe that is a cost we have to bear, and that overtime we will become better at doing so.

What makes it more challenging to write tests that pas against local and forked systems? An example would be useful

maurelian replied:
It is more challenging because devs cannot assume that: 

1. the system under test is fresh, so certain values become non-zero ([example fix diff](https://github.com/ethereum-optimism/optimism/commit/ed8e0db4455f7f40d6a0a294ec830060f7dcd48d#diff-0a1721cfd9ce4c8616f8e1ea94098f643518572bc8ead0605bafbee2255ff095R291-R293)).
2. the system was not necessarily created with the config in `hardhat.json`. Seeing reads to the `deploy.cfg()` thus becomes an antipattern (as [discussed here](https://github.com/ethereum-optimism/optimism/pull/13323#discussion_r1886934471)). Thus it becomes necessary to get config values from the chain state ([eg](https://github.com/ethereum-optimism/optimism/commit/3205090be2a1a3113fe004660532b0f95bded4d4#diff-47653a1adeb6dc8f4d4da44dadd3961ce110acf480bb1e03e7c8f8c82748b5bcR97-R101)).

Generally I think this is good as it will force us to write more robust and general tests, but it will take some adjustment.

maurelian replied:
I linked to my comment [here](https://github.com/ethereum-optimism/optimism/issues/13228) so that we can document these changes.

mds1 commented on packages/contracts-bedrock/justfile:
```diff
@@ -56,6 +56,37 @@ test *ARGS: build-go-ffi
+# - ETH_RPC_URL must be set to a production (Sepolia or Mainnet) RPC URL.
+# - FORK_BLOCK_NUMBER can be set in the env, or else will fallback to the default block number.
+#   Reusing the default block number greatly speeds up the test execution time by caching the
+#   rpc call responses in ~/.foundry/cache/rpc. The default block will need to be updated
+#   when the L1 chain is upgraded.
```
We should start a WIP runbook to document manual steps like this, so we don't forget them. Or this can be in a github issue, etc

maurelian replied:
Good call: https://github.com/ethereum-optimism/optimism/issues/13410

mds1 commented on packages/contracts-bedrock/test/setup/Upgrade.s.sol:
```diff
@@ -0,0 +1,114 @@
+
+// Interfaces
+import { IFaultDisputeGame } from "interfaces/dispute/IFaultDisputeGame.sol";
+import { IDisputeGame } from "interfaces/dispute/IDisputeGame.sol";
```
unused import

mds1 commented on packages/contracts-bedrock/test/setup/Upgrade.s.sol:
```diff
@@ -0,0 +1,114 @@
+    /// @notice Returns the OP chain name to use for forking
+    /// @return The OP chain name as a string
+    function upgradeOpChain() internal view returns (string memory) {
+        return vm.envOr("UPGRADE_OP_CHAIN", string("op"));
```
Do we have a tracking issue for a (perhaps scheduled) CI job that runs the upgrade tests for every base chain <> OP Chain pairing in the superchain registry, that we hold the keys for?

maurelian replied:
We do now! https://github.com/ethereum-optimism/optimism/issues/13401

mds1 commented on packages/contracts-bedrock/test/setup/CommonTest.sol:
```diff
@@ -42,17 +43,18 @@ contract CommonTest is Test, Setup, Events {
     IOptimismMintableERC20Full RemoteL1Token;
 
     function setUp() public virtual override {
+        // Because Setup.setup() may create a forked network, all cheatcodes must be run prior to any
+        // state modifying cheat codes.
```
nit: I think this is saying that state changes from state modifying cheats won't persist on the fork created by `Setup.setUp`? But it reads a bit confusingly, would be nice to clarify this

maurelian replied:
Yes, that's the intent but this was kind of a nonsense sentence. How is this?

```suggestion
        // Setup.setup() may switch the tests over to a newly forked network. Therefore
        // state modifying cheatcodes must be run after Setup.setup(), otherwise the 
        // changes will not be persisted into the new network. 
```

maurelian replied:
Making that change. 25573b8aa8e0

mds1 commented on packages/contracts-bedrock/test/setup/CommonTest.sol:
```diff
@@ -68,21 +70,27 @@ contract CommonTest is Test, Setup, Events {
+                vm.skip(true);
+            }
+
+            // Modifying these values on a fork test causes issues.
+            vm.warp(deploy.cfg().l2OutputOracleStartingTimestamp() + 1);
+            vm.roll(deploy.cfg().l2OutputOracleStartingBlockNumber() + 1);
+            vm.fee(1 gwei);
```
What issues does it cause? Also not following what this means since we are in a block where `isForkTest` is true, so are we causing the issues by modifying these values?

maurelian replied:
Yes, this diff fixed quite a few tests

Looking at the actual [deploy config values](https://github.com/ethereum-optimism/optimism/blob/upgrade-against-registry/packages/contracts-bedrock/deploy-config/hardhat.json#L17-L18) used in the test, this code block simplifies to: 

```
  vm.warp(2);
  vm.roll(2); 
  vm.fee(1 gwei);
```

This is quite disconnected from the actual state of the network, so we should expect that things break (ie. 2 step withdrawals and fault proving) when the global block number becomes significantly different from the block number which the contract state is based on.

maurelian replied:
Oh, facepalm. The logic here was wrong. I think I had it working then broke it, and given so many failing tests, it's easy to not notice that.
This is correct:

```
if (isForkTest) {
  ....
} else {
    // Modifying these values on a fork test causes issues.
    vm.warp(deploy.cfg().l2OutputOracleStartingTimestamp() + 1);
    vm.roll(deploy.cfg().l2OutputOracleStartingBlockNumber() + 1);
    vm.fee(1 gwei);
}
```

mds1 commented on packages/contracts-bedrock/test/setup/CommonTest.sol:
```diff
@@ -68,21 +70,27 @@ contract CommonTest is Test, Setup, Events {
+        }
+
         vm.etch(address(ffi), vm.getDeployedCode("FFIInterface.sol:FFIInterface"));
+        vm.allowCheatcodes(address(ffi));
```
Why did we need to add this?

maurelian replied:
Basically because it causes a bunch of tests to not fail. :)

For context, from the [foundry book](https://book.getfoundry.sh/cheatcodes/allow-cheatcodes):

> By default, the test contract, and its deployer are allowed to access cheatcodes. In addition to that, cheat code access is granted if the contract was deployed by an address that already has cheatcode access...

But the FfiInterface contract is not deployed by the test contract, it is etched directly into state [here](https://github.com/ethereum-optimism/optimism/blob/upgrade-against-registry/packages/contracts-bedrock/test/setup/CommonTest.sol#L34).

mds1 commented on packages/contracts-bedrock/test/setup/Setup.sol:
```diff
@@ -112,24 +117,60 @@ contract Setup {
-    ///      of this contract by fetching the bytecode dynamically using `vm.getCode()`.
-    ///      If the Deploy bytecode is included in this contract, then it will double
+    /// @dev Deploys either the Deploy.s.sol or Upgrade.s.sol contract, by fetching the bytecode dynamically using
+    ///      `vm.getCode()` and etching it into the state. This enables us to avoid including the bytecode of those
```
```suggestion
    ///      `vm.getDeployedCode()` and etching it into the state. This enables us to avoid including the bytecode of those
```

mds1 commented on packages/contracts-bedrock/test/setup/Upgrade.s.sol:
```diff
@@ -0,0 +1,114 @@
+        // After the upgrade is complete, the superchain-registry will be updated and the contract will be present.
+        // At this point, the test will need to be updated to read the new contract from the superchain-registry using
+        // either the `saveProxyAndImpl` or `save` functions.
+        string memory superchainToml =
+            vm.readFile(string.concat(superchainBasePath, upgradeBaseChain(), "/superchain.toml"));
+        string memory opToml =
+            vm.readFile(string.concat(superchainBasePath, upgradeBaseChain(), "/", upgradeOpChain(), ".toml"));
```
We may be able to simplify things by collapsing `UPGRADE_BASE_CHAIN` and `UPGRADE_OP_CHAIN` into `OP_CHAIN_ID` and reading all addresses from https://github.com/ethereum-optimism/superchain-registry/blob/main/superchain/extra/addresses/addresses.json

That is similar to what we are doing in superchain-ops: https://github.com/ethereum-optimism/superchain-ops/pull/406

maurelian replied:
I think it would be a significant refactor to this code path to:
- replace toml with json parsing
- needing to infer sepolia/mainnet from the data in the json file

And I'm not sure it gets us all that much. If this turns out not to scale well we can fix in the future. 
I'd like to resolve this without tracking it in an issue, please let me know if you feel strongly to the contrary.

mds1 reviewed (approved):
I like the architecture overall, nice work

ElliotFriedman commented on packages/contracts-bedrock/test/setup/Upgrade.s.sol:
```diff
@@ -0,0 +1,114 @@
+        IFaultDisputeGame permissionedDisputeGame =
+            IFaultDisputeGame(address(disputeGameFactory.gameImpls(GameTypes.PERMISSIONED_CANNON)));
+        save("PermissionedDisputeGame", address(permissionedDisputeGame));
+        save("PermissionedDelayedWETHProxy", address(permissionedDisputeGame.weth()));
```
if we imported Addresses.sol from FPS to manage contract addreses here, the devex might be better around managing addresses.

The contract calls could look like:

```
addresses.saveAddress("PermissionedDisputeGame", OP_MAINNET);
address addressManager = addresses.getAddress("AddressManager", OP_MAINNET);
```

maurelian replied:
I do agree that removing the Artifacts.sol utilities (ie. `save()` and `mustGetAdress()`) calls be for nice the purposes of just the solidity unit tests. In the simplest world we'd just directly save the deployment addresses to the [state vars](https://github.com/ethereum-optimism/optimism/blob/upgrade-against-registry/packages/contracts-bedrock/test/setup/Setup.sol#L80-L94), without going through another dep. 

But there are places outside of the foundry test suite that use the Artifacts.sol utilites ([eg](https://github.com/ethereum-optimism/optimism/blob/upgrade-against-registry/bedrock-devnet/devnet/__init__.py#L91)) to generate allocs for e2e testing which would make that a much bigger refactor than necessary to meet the objectives of this project.

maurelian committed:
feat: remove unused import

smartcontracts commented on packages/contracts-bedrock/test/L1/OptimismPortal.t.sol:
```diff
@@ -532,6 +532,8 @@ contract OptimismPortal_Test is CommonTest {
     /// @dev Tests that `depositERC20Transaction` reverts when the gas paying token is ether.
     function test_depositERC20Transaction_noCustomGasToken_reverts() external {
+        // TODO(opcm upgrades): remove skip once upgrade path is implemented
+        skipIfForkTest("OptimismPortal_Test: gas paying token functionality DNE on op mainnet");
```
Would be nice if `skipIfForkTest` took a chain ID as a parameter

maurelian replied:
Good call, though we'd need to either: 
1. take an array of chain IDs as a parameter. 
2. call `skipIfForkTest()` once per chain ID that its needed on.

Either way it's going to be ugly, so I'd like to defer it until we implement[ multichain CI testing](https://github.com/ethereum-optimism/optimism/issues/13401#issuecomment-2546282590).

smartcontracts replied:
Sgtm, happy to defer it

maurelian committed:
feat: fix test_findLatestGames_static_succeeds

maurelian committed:
fix: OptimismPortal2 test_initialize_succeeds

Read guardian from chain rather than config

maurelian committed:
feat: fix returnIfForkTest

Properly uses the return opcode to halt the test.

maurelian committed:
fix incorrect vm.roll/warp skipping logic

maurelian committed:
fix: Portal tests for forking

maurelian committed:
fix: DGF test for forking

maurelian committed:
fix: Skip CGT tests on L1StandardBridge

maurelian committed:
fix: Skip CGT tests on L1CrossDomainMessenger

maurelian committed:
fix DGF testFuzz_create_succeeds

maurelian committed:
fix: DGF testFuzz_create_sameUUID_reverts

maurelian committed:
feat: remove debugging annotations

maurelian committed:
fix: remove left over console

maurelian committed:
fix: OptimismPortal2 tests

maurelian committed:
fix: PermissionedDisputeGame tests

maurelian committed:
feat: Temporarily skip UnexpectedRootClaim failures on forked upgrade tests

maurelian committed:
fix gas snapshot

maurelian committed:
fix: remove unused import

maurelian committed:
fix: restore test_list to contracts-bedrock-tests job

maurelian committed:
feat(ci): Split up the test-upgrade CI job

maurelian committed:
fix(ci): remove erroneous param reference

maurelian committed:
fix: test_donateETH_succeeds for fork testing

maurelian committed:
feat: Improve caching of forked state

maurelian committed:
feat: use config syntax for inserting rpc url

maurelian committed:
debug: cache key

maurelian committed:
feat: Fix cache key checksum

maurelian committed:
Apply suggestions from code review

Co-authored-by: Matt Solomon <matt@mattsolomon.dev>

maurelian committed:
clarify comment about setup.setup forking

maurelian committed:
feat: remove diff meant for upstack

maurelian committed:
fix lint

maurelian committed:
Revert "feat: remove diff meant for upstack"

This reverts commit cab685617fd9db6961debe503f81042575b848f7.

maurelian committed:
feat: clearer env var name for NO_MATCH_CONTRACTS

blmalone commented on packages/contracts-bedrock/test/setup/Setup.sol:
```diff
@@ -112,24 +117,61 @@ contract Setup {
-    ///      of this contract by fetching the bytecode dynamically using `vm.getCode()`.
-    ///      If the Deploy bytecode is included in this contract, then it will double
+    /// @dev Deploys either the Deploy.s.sol or Upgrade.s.sol contract, by fetching the bytecode dynamically using
+    ///      `vm.getDeployedCode()` and etching it into the state. This enables us to avoid including the bytecode of
```
nit: fixed natspec formatting
```
    /// @dev Deploys either the Deploy.s.sol or Upgrade.s.sol contract, by fetching the bytecode dynamically using
    ///      `vm.getDeployedCode()` and etching it into the state.
    ///      This enables us to avoid including the bytecode of those contracts in the bytecode of this contract.
    ///      If the bytecode of those contracts was included in this contract, then it will double
    ///      the compile time and bloat all of the test contract artifacts since they
    ///      will also need to include the bytecode for the Deploy contract.
    ///      This is a hack as we are pushing solidity to the edge.

blmalone commented on packages/contracts-bedrock/test/L1/L1StandardBridge.t.sol:
```diff
@@ -211,7 +217,9 @@ contract PreBridgeETH is CommonTest {
     ///      on whether the bridge call is legacy or not.
     function _preBridgeETH(bool isLegacy, uint256 value) internal {
-        assertEq(address(optimismPortal).balance, 0);
+        if (!isForkTest) {
```
What do you think of making the `isForkTest` variable internal and only exposing it via a public getter? I'm thinking that in the future this could avoid some head scratching where someone accidentally sets `isForkTest` in their code and that has widespread effects on other tests. 

TLDR, I'm thinking the scope of setting the `isForkTest` variable should be limited to Setup.sol only. Then having a function like this to expose read only functionality to child classes: 

```
/// @dev Returns whether the setup is running in a forked production network.
/// @return True if running against a fork, false otherwise.
function getIsForkTest() public view returns (bool) {
    return isForkTest;
}
```

blmalone replied:
Additionally, following this approach might give some UX improvements in the tests. You could have two functions. 

```
function getIsGenesisTest() public view returns (bool) {
    return !isForkTest;
}

function getIsForkTest() public view returns (bool) {
    return isForkTest;
}
```

maurelian replied:
I like this idea a lot, though will only do the one getter. It can be negated with `!getIsForkTest()` if necessary.

maurelian replied:
b02c91c88ee6

blmalone commented on packages/contracts-bedrock/test/L1/OptimismPortal.t.sol:
```diff
@@ -1338,7 +1340,8 @@ contract OptimismPortal_FinalizeWithdrawal_Test is CommonTest {
     /// @dev Tests that the proxy is initialized correctly.
-    function test_params_initValuesOnProxy_succeeds() external view {
+    function test_params_initValuesOnProxy_succeeds() external {
+        skipIfForkTest("OptimismPortal_Test: resource config varies on mainnet");
```
nit: not sure if you want this comment in here too or not. 
```suggestion
        // TODO(opcm upgrades): remove skip once upgrade path is implemented
        skipIfForkTest("OptimismPortal_Test: resource config varies on mainnet");
```

maurelian replied:
No, The problem here is that the values on mainnet are very dynamic updating with every deposit tx, so they just don't match the initial values anymore.

blmalone commented on packages/contracts-bedrock/test/L1/OptimismPortal2.t.sol:
```diff
@@ -134,6 +141,9 @@ contract OptimismPortal2_Test is CommonTest {
     /// @dev Tests that `receive` successdully deposits ETH.
     function testFuzz_receive_succeeds(uint256 _value) external {
+        uint256 balanceBefore = address(optimismPortal2).balance;
+        _value = bound(_value, 0, type(uint256).max - balanceBefore);
```
Nice! It's now important for devs to remember that all tests are either ran from a brand new deploy or a forked chain. It informs how the tests are written. This is a good example of such a test.

maurelian replied:
Yep, this becomes a common pattern!

blmalone commented on packages/contracts-bedrock/test/L1/OptimismPortal2.t.sol:
```diff
@@ -1607,7 +1642,8 @@ contract OptimismPortal2_Upgradeable_Test is CommonTest {
     /// @dev Tests that the proxy is initialized correctly.
-    function test_params_initValuesOnProxy_succeeds() external view {
+    function test_params_initValuesOnProxy_succeeds() external {
+        skipIfForkTest("OptimismPortal2_Test: resource config varies on mainnet");
```
```suggestion
        // TODO(opcm upgrades): remove skip once upgrade path is implemented
        skipIfForkTest("OptimismPortal2_Test: resource config varies on mainnet");
```

maurelian replied:
Same, init values on hardhat.json differ from mainnet.

blmalone commented on packages/contracts-bedrock/test/L1/ProtocolVersions.t.sol:
```diff
@@ -24,7 +24,10 @@ contract ProtocolVersions_Init is CommonTest {
     /// @dev Tests that initialization sets the correct values.
-    function test_initialize_values_succeeds() external view {
+    function test_initialize_values_succeeds() external {
+        skipIfForkTest(
```
```suggestion
        // TODO(opcm upgrades): remove skip once upgrade path is implemented
        skipIfForkTest(
```

maurelian replied:
No, the issue here is that the config differs of op-mainnet differs from the hardhat.json config. So we don't really have a good way of testing this.

blmalone commented on packages/contracts-bedrock/test/L1/SuperchainConfig.t.sol:
```diff
@@ -12,6 +12,11 @@ import { ISuperchainConfig } from "interfaces/L1/ISuperchainConfig.sol";
 contract SuperchainConfig_Init_Test is CommonTest {
+    function setUp() public virtual override {
+        super.setUp();
+        skipIfForkTest("SuperchainConfig_Init_Test: cannot test initialization on forked network");
```
```suggestion
        // TODO(opcm upgrades): remove skip once upgrade path is implemented
        skipIfForkTest("SuperchainConfig_Init_Test: cannot test initialization on forked network");
```

maurelian replied:
same

blmalone commented on packages/contracts-bedrock/test/dispute/DisputeGameFactory.t.sol:
```diff
@@ -172,7 +172,9 @@ contract DisputeGameFactory_SetInitBond_Test is DisputeGameFactory_Init {
     function test_setInitBond_succeeds() public {
         // There should be no init bond for the `GameTypes.CANNON` enum value, it has not been set.
-        assertEq(disputeGameFactory.initBonds(GameTypes.CANNON), 0);
+        if (!isForkTest) {
```
Similar comment as I left above. Might be a good idea to use a public getter here.

maurelian replied:
b02c91c88ee6

blmalone commented on packages/contracts-bedrock/test/dispute/FaultDisputeGame.t.sol:
```diff
@@ -48,7 +48,9 @@ contract FaultDisputeGame_Init is DisputeGameFactory_Init {
     function init(Claim rootClaim, Claim absolutePrestate, uint256 l2BlockNumber) public {
         // Set the time to a realistic date.
-        vm.warp(1690906994);
+        if (!isForkTest) {
```
again

maurelian replied:
b02c91c88ee6

blmalone commented on packages/contracts-bedrock/test/setup/Upgrade.s.sol:
```diff
@@ -0,0 +1,113 @@
+        // Superchain shared contracts
+        saveProxyAndImpl("SuperchainConfig", superchainToml, ".superchain_config_addr");
+        saveProxyAndImpl("ProtocolVersions", superchainToml, ".protocol_versions_addr");
+        save("OPContractsManager", vm.parseTomlAddress(superchainToml, ".op_contracts_manager_proxy_addr"));
```
I would like to double check that the SCR is storing the correct OPCM addresses. I know there's been some churn and the canonical reference at this point is likely in op-deployer: https://github.com/ethereum-optimism/optimism/blob/develop/op-deployer/pkg/deployer/standard/standard.go#L202-L215

The OPCM currently stored in SCR is still proxied: https://etherscan.io/address/0x18CeC91779995AD14c880e4095456B9147160790#code 

More broadly, I think we need to all agree on a format that we want to store OPCM addresses in SCR. I've created an issue for this in SCR. I'm not sure where best to make sure it's being tracked. It's an important part of this work though: https://github.com/ethereum-optimism/superchain-registry/issues/806

maurelian replied:
Good call, I've also noticed that.

I added it to my [master issue](https://github.com/ethereum-optimism/optimism/issues/13069) under Milestone:Mainnet.

blmalone reviewed (approved):
This is look good. Admittedly I'm not as familiar with the ins and outs of the circleci job though it looks sensible. 

I see that CI is currently failing but I'm guessing the fix includes skipping more tests that cannot be run on a forked chain. 

Approving.

maurelian committed:
feat: make isForkTest private and add a getter

maurelian committed:
fix: isForkTest getter call

maurelian committed:
fix gas snapshot

maurelian committed:
Merge pull request #13424 from ethereum-optimism/upgrade-against-registry-2

Fix tests on forked networks

maurelian committed:
fix test_constructor_succeeds tests after disableInit rebase

maurelian committed:
lint

maurelian committed:
Merge pull request #13428 from ethereum-optimism/upgrade-against-registry-3

Skip UnexpectedRootClaim failures

maurelian commented on .circleci/config.yml:
```diff
@@ -1195,7 +1256,7 @@ workflows:
       - contracts-bedrock-tests:
           # Test everything except PreimageOracle.t.sol since it's slow.
           name: contracts-bedrock-tests
-          test_list: find test -name "*.t.sol" -not -name "PreimageOracle.t.sol"
```
I introduced a bug here which caused ONLY the preimage oracle tests to run.
